pkgs/applications/emulators/retroarch/wrapper.nix

```diff
@@ -3,16 +3,28 @@
 , makeWrapper
 , retroarch
 , symlinkJoin
+, runCommand
 , cores ? [ ]
+, settings ? { }
 }:
 
 let
+  settingsPath = runCommand "declarative-retroarch.cfg"
+    {
+      value = lib.concatStringsSep "\n" (lib.mapAttrsToList (n: v: "${n} = \"${v}\"") settings);
+      passAsFile = [ "value" ];
+    }
+    ''
+      cp "$valuePath" "$out"
+    '';
```

Comment:
I'm happy I found it. I started patching retroarch to add support for build-time injecting retroarch-joypad-autoconfig, and then realized it could be done in the wrapper instead, without patching. This also means we can remove our existing nixpkgs specific patch. But I didn't want to do that in this PR, because I wanted to make this one backward compatible.

Reply:
@bjornfor If it is possible we should just remove the current patches. I am not completely convinced because I remember trying a few of the flags in RetroArch and they did not work (they're not updated when the path changed when e.g. a new release), but I am not sure if I tried `--appendconfig`.

I would say you should try to submit the changes anyway, because as long it works as expected they should be backwards compatible (since we patch the code directly and would stop doing it, the change in functionality would be none).

Reply:
@thiagokokada:
> [...] I remember trying a few of the flags in RetroArch and they did not work (they're not updated when the path changed when e.g. a new release), but I am not sure if I tried --appendconfig.

--appendconfig= seems to win over $HOME/.config/retroarch/retroarch.cfg and internal built-in settings. And settings in it get updated/reflected back into retroarch.cfg. So retroarch.cfg still represents the truth, but it might get programmatically updated when retroarch is started with a new --appendconfig=.

> I would say you should try to submit the changes anyway, [...]

 I'm not very enthusiastic/motivated now that I know about `--appendconfig=`. The `--with-joypad_autoconfig_dir=` patch I was testing was basically copy paste from your `--with-assert_dir=` patch (https://github.com/libretro/RetroArch/commit/894c44c5ea7f1eada9207be3c29e8d5c0a7a9e1f). And having to patch upstream for each path/setting we'd like to configure at build-time seems like a bad idea.

What I would consider though (but very low on my TODO list) is letting users own ~/.config/retroarch/retroarch.cfg. Currently it's a weird mix of application and user owned data, as the application dumps all its state/settings there, and then never update it again (leaving the ownership/maintenance to users). IMHO that file should be empty by default and only settings the user explicitly wanted to change should be stored there. At that point, one could share retroarch.cfg across systems / package managers.

Reply:
> --appendconfig= seems to win over $HOME/.config/retroarch/retroarch.cfg and internal built-in settings. And settings in it get updated/reflected back into retroarch.cfg. So retroarch.cfg still represents the truth, but it might get programmatically updated when retroarch is started with a new --appendconfig=.

If what you're saying is true (that `--appendconfig` has priority over user config), I think we can drop [this patch](https://github.com/NixOS/nixpkgs/blob/master/pkgs/applications/emulators/retroarch/use-default-values-for-libretro_info_path-assets_directory.patch) and the `--with-assert_dir`. This would be useful already.

If you could do this I would be grateful, otherwise I will probably leave this at my TODO list when the next version of RetroArch comes.

> IMHO that file should be empty by default and only settings the user explicitly wanted to change should be stored there. At that point, one could share retroarch.cfg across systems / package managers.

I think the easiest way to do so is to create a module (either a NixOS or Home-Manager module, for NixOS I think RetroArch supports configs in `/etc/retroarch/retroarch.conf` too). Using either of those will create a read-only file that should be immutable.

I hope RetroArch itself fails gracefully once it can't write the config file, but if it doesn't I think this would be a patch upstream would easily accept.

Reply:
> I think we can drop [this patch](https://github.com/NixOS/nixpkgs/blob/master/pkgs/applications/emulators/retroarch/use-default-values-for-libretro_info_path-assets_directory.patch) and the --with-assert_dir/--with-core_info_dir.

Yes.

But I'm not sure what to do about the corresponding `--disable-update_assets` and `--disable-update_core_info` options. Always pass them or remove them?

(It looks like RetroArch has `--enable-*/--disable-*` options for configure time configuration, and a somewhat different set of options for runtime configuration in retroarch.cfg. Some overlap exists, like the path settings, but I don't think we can build the wrapper with `settings = { "disable-update_assets" = "true"; }` and expect it to work.)

> I think the easiest way to do so is to create a module (either a NixOS or Home-Manager module [...]

I don't think that's a good idea -- it makes us different and relies on corner case behaviour: how does RetroArch handle an immutable config file?

I want this fixed upstream: retroarch should only write stuff to ~/.config/retroarch/retroarch.cfg when it *differs from the built-in config*. I think it's fine for the file to be managed imperatively -- either by user actions in $EDITOR or RetroArch UI -- as long as it only contains the settings the user has changed and not a full state dump from the application.

Reply:
> But I'm not sure what to do about the corresponding --disable-update_assets and --disable-update_core_info options. Always pass them or remove them?

Keep them as build options.

> I don't think that's a good idea -- it makes us different and relies on corner case behaviour: how does RetroArch handle an immutable config file?

It is basically the whole idea of NixOS and Home-Manager modules, they always relies on how the program handle an read-only config file. This is not supposed to be a corner case behaviour, every program needs to handle when the file is not writable correctly.

Reply:
Ok, I guess you're right about read-only config being common. But on the other hand, if we use git as an example, it does not make changes to ~/.config/git/config unless the user asks for it. And it doesn't dump all internal config there either (which can get stale on next nixpkgs update).

I tested with a root owned (and minimal/empty) ~USER/.config/retroarch/retroarch.cfg and RetroArch says `Failed saving config to "/home/USER/.config/retroarch/retroarch.cfg"` (in the UI) but otherwise runs fine. When I change a setting in the UI, it doesn't say anything (it allows changes) but when I restart RetroArch it's clear that the settings weren't persisted.

IOW, home-manager style config is doable. But then the user must be aware that even though the UI gives the impression of allowing changes (there's no warning against it after the startup warning -- which quickly goes away), it doesn't work.

It would be nice if upstream would show a warning about being unable to save the config on each attempted config change, and not just at startup. (Saving the config at startup is silly -- the user didn't make any configuration changes yet.)

Note that, although a home-manager style config is doable, having to block an app from dumping state in $HOME to prevent it from breaking itself on the next nixpkgs update is wrong.

Reply:
> Ok, I guess you're right about read-only config being common. But on the other hand, if we use git as an example, it does not make changes to ~/.config/git/config unless the user asks for it. And it doesn't dump all internal config there either (which can get stale on next nixpkgs update).

```
$ git config --global user.signingKey true
error: could not lock config file /Users/thiagoko/.config/git/config: Permission denied
```

I can understand your point that git is only changing when I am explicitly asking it to change my config, but the fact is that the whole idea of NixOS and Home-Manager modules is that we are going probably breaking some features in the program (like in this case `git config`) to get reproducible environments. That is the whole idea.

> Note that, although a home-manager style config is doable, having to block an app from dumping state in $HOME to prevent it from breaking itself on the next nixpkgs update is wrong.

Sure, I never said that what upstream is doing is reasonable and I would love to see this being fixed.